kevinpapst/kimai2
# add ldap support

## Description

Documentation can be found at: https://www.kimai.org/documentation/ldap.html

Features:
- Authenticate against an LDAP
- New users will be created on the fly
- Configured attributes will be synced on every login
- Mapping LDAP groups to Kimai roles is possible

**TODO**
- [x] prevent setting an internal password
- [x] get rid of the service class name parameter
- [x] convert ldap container parameter to config class
- [x] add group mapping query

**How to test**
- checkout the branch
- `composer install`
- adjust `config/packages/local.yaml` to your needs (see [documentation](https://www.kimai.org/documentation/ldap.html))
- `bin/console cache:clear`

**Open questions for EVERY reader**

- How do other apps handle LDAP support?
- Should an app still offer registration when LDAP is activated? 
  - Leave that configurable or turn it off always?
  - I guess we need to avoid problems like a local account with the same account name in LDAP, then the LDAP account would never be checked?
- Should groups be transferred?
  - How? Is there a general approach?
  - I read something about a "memberof" field/attribute which seems to only a OpenLDAP feature?
- The login needs to be checked against LDAP every time?
- What should happen to the local account if an account in LDAP is deleted?
- Can LDAP accounts be deactivated?

If anything else comes to your mind, please share it. This is one of the features I have to develop without having a proper LDAP setup/knowledge, so I am "flying blindly" here and need YOUR input. 

Fixes #193 
Fixes #795
Fixes #449 

## Types of changes

## Checklist
- [x] I verified that my code applies to the guidelines (`composer code-check`)
- [x] I updated the documentation (see [here](https://github.com/kimai/www.kimai.org/tree/master/_documentation))
- [x] I agree that this code is used in Kimai and will be published under the [MIT license](https://github.com/kevinpapst/kimai2/blob/master/LICENSE)
